langmeta/langmeta/shared/src/main/scala/org/langmeta/semanticdb/internal/package.scala

```diff
@@ -10,132 +10,192 @@ import org.langmeta.inputs.{Input => dInput}
+        sdocuments
       } else {
-        sdatabase.documents match {
-          case Seq(doc, messages)
-            if doc.filename == messages.filename &&
-                isOnlyMessages(messages) =>
-            val x = doc.addMessages(messages.messages: _*)
-            s.Database(x :: Nil)
-          case _ => sdatabase
+        sdocuments.documents match {
+          case Seq(smaindoc, sdiagdoc)
+            if smaindoc.uri == sdiagdoc.uri &&
+                isOnlyMessages(sdiagdoc) =>
+            val smaindoc1 = smaindoc.addDiagnostics(sdiagdoc.diagnostics: _*)
+            s.TextDocuments(smaindoc1 :: Nil)
+          case _ => sdocuments
         }
       }
     }
     def toVfs(targetroot: AbsolutePath): v.Database = {
-      val ventries = sdatabase.documents.toIterator.map { sentry =>
+      val ventries = sdocuments.documents.toIterator.map { sentry =>
         // TODO: Would it make sense to support multiclasspaths?
         // One use-case for this would be in-place updates of semanticdb files.
-        val vpath = v.SemanticdbPaths.fromScala(RelativePath(sentry.filename))
+        val vpath = v.SemanticdbPaths.fromScala(RelativePath(sentry.uri))
         val fragment = Fragment(targetroot, vpath)
-        val bytes = s.Database(List(sentry)).toByteArray
+        val bytes = s.TextDocuments(List(sentry)).toByteArray
         v.Entry.InMemory(fragment, bytes)
       }
       v.Database(ventries.toList)
     }
 
-    def toDb(sourcepath: Option[Sourcepath], sdoc: s.Document): d.Document = {
-      val s.Document(sunixfilename, scontents, slanguage, snames, smessages, ssymbols, ssynthetics) = sdoc
-      assert(sunixfilename.nonEmpty, "s.Document.filename must not be empty")
-      val sfilename = PathIO.fromUnix(sunixfilename)
+    def toDb(sourcepath: Option[Sourcepath], sdoc: s.TextDocument): d.Document = {
+      val s.TextDocument(sformat, suri, stext, slanguage, ssymbols, soccurrences, sdiagnostics, ssynthetics) = sdoc
+      assert(sformat == "semanticdb3", "s.TextDocument.format must be \"semanticdb2\"")
       val dinput = {
-        if (scontents == "") {
-          val uri =
+        val sfilename = {
+          assert(suri.nonEmpty, "s.TextDocument.uri must not be empty")
+          PathIO.fromUnix(suri)
+        }
+        if (stext == "") {
+          val duri =
             sourcepath.getOrElse(sys.error("Sourcepath is required to load slim semanticdb."))
                 .find(RelativePath(sfilename))
                 .getOrElse(sys.error(s"can't find $sfilename in $sourcepath"))
-          dInput.File(AbsolutePath(uri.getPath))
+          dInput.File(AbsolutePath(duri.getPath))
         } else {
-          dInput.VirtualFile(sfilename.toString, scontents)
+          dInput.VirtualFile(sfilename.toString, stext)
+        }
+      }
+      object sRange {
+        def unapply(srange: s.Range): Option[dPosition] = {
+          val dstartOffset = dinput.lineToOffset(srange.startLine) + srange.startCharacter
```

Comment:
we should make lineToOffset non-package-private, both metals and metaconfig have custom files in org.langmeta package to access this method. Now that line/character are the default I think we should bake these into the public API.

Maybe even add new Position.Range constructor: `Position.Range.fromLine(input, startLine, startColumn, endLine, endColumn)`

Reply:
Done.